src/storage/src/source/postgres.rs

```diff
@@ -428,12 +428,25 @@ async fn postgres_replication_loop_inner(
                 );
             }
             Err(ReplicationError::Indefinite(e)) => {
+                // If the channel is shutting down, so is the source.
+                let _ = task_info
+                    .sender
+                    .send(InternalMessage::Status(SourceStatusUpdate {
+                        status: SourceStatus::Stalled,
+                        error: Some(e.to_string()),
+                    }))
+                    .await;
                 // TODO: In the future we probably want to handle this more gracefully,
                 // but for now halting is the easiest way to dump the data in the pipe.
                 // The restarted storaged instance will restart the snapshot fresh, which will
                 // avoid any inconsistencies. Note that if the same lsn is chosen in the
                 // next snapshotting, the remapped timestamp chosen will be the same for
                 // both instances of storaged.
+                // In the meantime, we'd like to increase the changes that the stall reported
+                // above will make it to the status history table, so we'll wait a little
+                // bit before halting. This should be removed once we have a more graceful
+                // shutdown approach.
+                tokio::time::sleep(Duration::from_secs(10)).await;
```

Comment:
aha! I see what you mean: the path that the above status update takes is quite long, it goes out to the thing that drives the `SourceReader`, from there goes across timely channels to a different operator, and only there is it written into the status shard. And all that has to happen before we `halt!` here because that stops the whole process.

Though one... :see_no_evil:

Strawman: write to the status shard right here, before doing the `halt!`?

Reply:
Can't do it I think! We don't have access to persist here, and it looks like it would be a challenge to thread all the way through.

If we're nervous about the sleep -- and I think that's a reasonable stance -- I should probably just revert the commit and eat the slightly worse error reporting.